### PR TITLE
[webgpu] Use i32 as dimension type in WGSL

### PR DESCRIPTION
**Pull request by axinging**
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5602)

### CODE REVIEW

**Comment by axinging** [comment]
@qjia7 @haoyunfeix @xhcao @gyagp , PTAL.

File: tfjs-backend-webgpu/src/kernels/strided_slice_webgpu.ts [hunk 1]
```diff
@@ -91,17 +91,17 @@ export class StridedSliceProgram implements WebGPUProgram {
...
       let outputAxis = 0;
       newCoords =
           this.outputShape
               .map((_, i) => {
                 outputAxis++;
                 return this.outputShape.length === 1 ?
-                    `coords * uniforms.strides[${i}] + uniforms.begin[${i}]` :
-                    `coords[${outputAxis - 1}] * uniforms.strides[${
-                        i}] + uniforms.begin[${i}]`;
+                    `coords * u32(uniforms.strides[${i}]) + u32(uniforms.begin[${i}])` :
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/strided_slice_webgpu.ts:102]: For https://js.tensorflow.org/api/latest/#stridedSlice, we can see that `a stride can be negative, which causes a reverse slice.`. But here you are converting it to `u32`. Is it a problem?

File: tfjs-backend-webgpu/src/kernels/slice_webgpu.ts [hunk 2]
```diff
@@ -47,7 +47,7 @@ export class SliceProgram implements WebGPUProgram {
 
     this.start = start;
     this.uniforms = `${getCoordsDataType(start.length)} start; `;
-    this.uniformsWgsl = `start : ${getCoordsDataTypeWgsl(start.length)}; `;
+    this.uniformsWgsl = `start : ${getCoordsDataTypeWgsl(start.length, 'i32')}; `;
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/slice_webgpu.ts:50]: It seems that u32 is better.

File: tfjs-backend-webgpu/src/kernels/reduce_webgpu.ts [hunk 3]
```diff
@@ -30,7 +30,7 @@ export class ReduceProgram implements WebGPUProgram {
   workGroupSize: [number, number, number];
   variableNames = ['x'];
   uniforms = 'int reduceSize;';
-  uniformsWgsl = 'reduceSize : u32;';
+  uniformsWgsl = 'reduceSize : i32;';
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/reduce_webgpu.ts:33]: It's not necessary to change `reduceSize` type from `u32` to `i32`？
And it's also weird to do below change:
```
       fn DIV_CEIL(a : u32, b : u32) -> u32 {
        return u32(i32(a) - 1) / b + 1u;
       }
```
if a = 0u, in anyway, it is not correct unless you change a/b's type to `i32`

File: tfjs-backend-webgpu/src/kernels/pad_webgpu.ts [hunk 4]
```diff
@@ -99,10 +99,11 @@ export class PadProgram implements WebGPUProgram {
     const rank = this.xShape.length;
     const type = getCoordsDataTypeWgsl(rank);
     // The length of paddings are same with the rank of the input tensor.
-    const start = this.xShape.map((_, i) => `uniforms.pad${i}[0]`).join(',');
+    const start =
+        this.xShape.map((_, i) => `u32(uniforms.pad${i}[0])`).join(',');
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/pad_webgpu.ts:103]: Is it possible that `pad[i]` is a negative value? But you are converting it to `u32`. Can you add a unit test for this kind of situation?

File: tfjs-backend-webgpu/src/kernels/matmul_packed_vec4_webgpu.ts [hunk 5]
```diff
@@ -194,7 +194,7 @@ export function makeMatMulPackedVec4SourceWgsl(
 
     let globalRow = globalId.y * RowPerThread;
     let globalCol = globalId.x;
-    let numTiles = (uniforms.dimInner - 1u) / TileInner + 1u;
+    let numTiles = u32(i32(uniforms.dimInner) - 1) / TileInner + 1u;
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/matmul_packed_vec4_webgpu.ts:197]: Is it better to use `let numTiles = (i32(uniforms.dimInner) - 1) / ${tileInner} + 1;`? It's just error-prone to change a i32 to u32.

**Comment by qjia7** [comment]
@pyu10055 @lina I have a bad feeling about the current change to mix `u32` and `i32` together. It's error-prone and not good for maintenance. Maybe it's time to reconsider that whether we should unify all types to `i32`. What's your opinion on this?

File: tfjs-backend-webgpu/src/kernels/argminmax_webgpu.ts [hunk 6]
```diff
@@ -31,7 +31,7 @@ export class ArgMinMaxProgram implements WebGPUProgram {
   workGroupSize: [number, number, number];
   variableNames = ['x'];
   uniforms = 'int axis;';
-  uniformsWgsl = 'axis : u32;';
+  uniformsWgsl = 'axis : i32;';
```
[the remark this answers is not in the document]
Comment (pyu10055) [reply tfjs-backend-webgpu/src/kernels/argminmax_webgpu.ts:34]: the axis could be between -rank(input), rank(input), but tfjs-core should handle the conversion of the axis to unsigned value.

**Review by lina128 — COMMENTED** [review]
Along your suggestion Jiajia, I wonder if we should also use i32 for shapes too? Basically eliminate u32 everywhere. I checked TF's TensorShape class, they seem to use i32 for dimensions. 

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5602#-:-Mj_IOfy1lU-zL1VUpl-:b-rpd9qr)* status: 0 of 1 approvals obtained (waiting on @axinging, @pyu10055, and @qjia7)

**Review by pyu10055 — COMMENTED** [review]
are there any performance difference between i32 and u32? 
I think u32 would be simpler for the shader, since the branching logic ( for values < 0 ) should be taken cared in js than the shader.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5602#-:-Mj_Jzsw0w8WLpQDkf4e:b-63u55y)* status: 0 of 1 approvals obtained (waiting on @axinging and @qjia7)

**Review by lina128 — COMMENTED** [review]
I believe that's the original motivation to use u32. But then it can cause hard-to-debug bugs because of the implicit conversion for values < 0, in some intermediate steps, dimension can be negative, we still need to explicitly check the value at the end and set to 0 for negative values. Also Jiajia said the performance is the same between i32 and u32.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5602#-:-Mj_MIkYAe4810Zd64_B:b-rhd60e)* status: 0 of 1 approvals obtained (waiting on @axinging and @qjia7)

**Comment by qjia7** [comment]
@axinging As we discussed in the meeting, we all agree to use i32 in all places, including the shapes, coords and uniforms. Please update this PR to follow this rule. Thanks.

**Pull request closed without merging**